migrations/back-37/back-37.go

```diff
@@ -0,0 +1,202 @@
+	dataRepository *storeStructuredMongo.Repository
+}
+
+func NewMigration() *Migration {
+	return &Migration{
+		Migration: migrationMongo.NewMigration(),
+	}
+}
+
+func (m *Migration) Initialize(provider application.Provider) error {
+	if err := m.Migration.Initialize(provider); err != nil {
+		return err
+	}
+
+	m.CLI().Usage = "BACK-37: Migrate all existing data to add required Platform deduplication hash fields"
+	m.CLI().Description = "BACK-37: To fully migrate devices from the `jellyfish` upload API to the `platform` upload API"
+	m.CLI().Authors = []cli.Author{
+		{
+			Name:  "J H BATE",
+			Email: "jamie@tidepool.org",
+		},
+	}
+
+	m.CLI().Action = func(ctx *cli.Context) error {
+		if !m.ParseContext(ctx) {
+			return nil
+		}
+		return m.execute()
+	}
+
+	return nil
+}
+
+func (m *Migration) execute() error {
+	m.Logger().Debug("Migrate jellyfish upload API data")
+	m.Logger().Debug("Creating data store")
+
+	mongoConfig := m.NewMongoConfig()
+	mongoConfig.Database = "data"
+	mongoConfig.Timeout = 60 * time.Minute
+	dataStore, err := storeStructuredMongo.NewStore(mongoConfig)
+	if err != nil {
+		return errors.Wrap(err, "unable to create data store")
+	}
+	defer dataStore.Terminate(context.Background())
+
+	m.Logger().Debug("Creating data repository")
+
+	m.dataRepository = dataStore.GetRepository("deviceData")
+	hashUpdatedCount, archivedCount, errorCount := m.migrateJellyfishDocuments()
+	m.Logger().Infof("Migrated %d jellyfish documents", hashUpdatedCount)
+	m.Logger().Infof("Archived %d duplicate jellyfish documents", archivedCount)
+	m.Logger().Infof("%d errors occurred", errorCount)
+
+	return nil
+}
+
+func (m *Migration) migrateJellyfishDocuments() (int, int, int) {
+	logger := m.Logger()
+
+	logger.Debug("Finding distinct users")
+
+	var hashUpdatedCount, archivedCount, errorCount int
+
+	userIDs, err := m.dataRepository.Distinct(context.Background(), "_userId", bson.M{})
+	if err != nil {
+		logger.WithError(err).Error("Unable to execute distinct query")
+	} else {
+		logger.Debugf("Finding jellyfish records for %d users", len(userIDs))
+
+		for _, userID := range userIDs {
+			logger.Debugf("Finding jellyfish records for user ID %s", userID)
+			selector := bson.M{
+				"_userId": userID,
+				"_active": true,
+				// uploads aren't de-duped.
+				"type":          bson.M{"$ne": "upload"},
+				"_deduplicator": bson.M{"$exists": false},
+			}
+
+			var jellyfishResult bson.M
+			jellyfishDocCursor, err := m.dataRepository.Find(context.Background(), selector)
+			if err != nil {
+				logger.WithError(err).Error("Unable to find jellyfish results")
+			}
+			for jellyfishDocCursor.Next(context.Background()) {
+				jellyfishDocCursor.Decode(&jellyfishResult)
+
+				dupQuery := bson.M{
+					"_userId":       userID,
+					"_active":       true,
+					"_id":           jellyfishResult["_id"],
+					"_deduplicator": bson.M{"$exists": true},
+				}
+				dupCursor, err := m.dataRepository.Find(context.Background(), dupQuery)
```

Comment:
would this ever return a document since the  *_id* is unique and the original query on line 101 already checks for no dedeupe hash? Or can the doc be modified in b/t this query and the one on line 105?

Reply:
the `_id` is yet another hash of the datum's "id" fields e.g. for the smbg `var idFields = ['type', 'deviceId', 'time', 'value'];`

see https://github.com/tidepool-org/jellyfish/blob/35165c8dce520390213fd2fbd7b0228931d976e3/lib/schema/smbg.js#L22 for example 

So if you had a duplicate value they would have the same `_id` from what I understand? 

@darinkrauss ??

Reply:
There are two fields, the internal mongodb `_id` and our own generated `id` I believe. I think you want the `id` here.